Fb365

joelaro04 opened the pull request.

joelaro04 committed:
Merge branch 'master' of github.com:spiretrading/pyaugur

joelaro04 committed:
Implemented base AugurClient class.

joelaro04 committed:
Implementation for load_market_info method

joelaro04 committed:
Cast MarketInfo parameters to Decimal and datetime

joelaro04 committed:
Added addresses.json

joelaro04 committed:
Added exception handler for AugurClient

joelaro04 committed:
Merged with master

joelaro04 committed:
Translated camel case from node to snake case

joelaro04 committed:
Removed requests import

joelaro04 committed:
Removed unused variables

joelaro04 committed:
Cast reporting_state from node to a ReportingState enum

kranar commented on augur/market_info.py:
```diff
@@ -1,4 +1,9 @@
+
+from datetime import datetime
+from decimal import Decimal
+
```
This file should not change. Only the client should be modified.

joelaro04 replied:
I changed some property names to match the data we get from the node.

joelaro04 committed:
Dynamically cast types for MarketInfo objects

joelaro04 committed:
Reset MarketInfo

kranar commented on augur/augur_client.py:
```diff
@@ -0,0 +1,134 @@
+    for key, value in response_data.items():
+      new_key = self._to_snake_case(key)
+      if value is None:
+        market_info_dict[new_key] = None
```
This won't work in cases where an empty list or str is used instead of None.

joelaro04 committed:
MarketInfo.tags does not return None. It can return a list of None.

joelaro04 committed:
Cast OutcomeInfo and added validation for empty list/strings

joelaro04 committed:
Omit all None types in lists.

- Fixed datetime.fromtimestamp from raising and error when passed
microseconds.

joelaro04 committed:
Removed MarketInfo.tags returning a list of None.

joelaro04 committed:
Properties that expect a list or str will no longer return None

joelaro04 committed:
Added empty list/default for OutcomeInfo and NormalizedPayout in AugurClient.

joelaro04 committed:
Removed shortnames

kranar commented on augur/augur_client.py:
```diff
@@ -46,50 +46,73 @@ def __init__(self, hostname, port):
-          new_value = [OutcomeInfo(**x) for x in value]
+      to_cast_type = casting_map.get(new_key, None)
+      if not value:
+        if to_cast_type is list:
```
This appears to be incorrect. Line 88 checks if ```value``` is ```False``` and if it is then...

1. Line 90 will iterate over all elements in ```value```, but ```value``` is ```False``` which means that it's either ```None``` or an empty ```list```, so iterating it will have no effect.

2. Line 92 will assign an empty string if ```not value```, but you already know that ```not value``` is ```True``` since you checked for that on line 88.

3. Line 94 assigns a result of ```None```, but what if ```value``` is of type ```bool``` and is ```False```, then you will assign a value of ```None``` to ```new_value``` when you should have assigned ```False```.

My suspicion is that the approach of having a giant dictionary with values with a callable object used to cast the value might have seemed like a good idea initially to save code, that it's becoming too error prone to use this method in practice.

I suspect that implementing this using the more direct approach will end up in less code written and code that is more easy to verify and reason about, ie:

```
  return MarketInfo(response_data['x'], Decimal(response_data['y']), datetime(response_data['z']), ...)
```

joelaro04 committed:
Cast Augur types

kranar commented on augur/augur_client.py:
```diff
@@ -44,76 +44,105 @@ def __init__(self, hostname, port):
+    if response_data['lastTradeTime']:
+      response_data['lastTradeTime'] = (
+        self._ensure_unix_timestamp(response_data['lastTradeTime']))
+    # Ensure decimals default to None if needed.
```
Can you explain what you mean by defaulting to ```None``` if needed? ```None``` is not a valid ```Decimal``` and will raise an exception.

joelaro04 replied:
It looks like the default for `initialReportSize`, `finalizeBlockNumber`, `lastTradeBlockNumber` is already `None`, so lines 60-68 isn't needed. I will need to ensure that if we do get None that it doesn't get cast to a `Decimal`.

kranar commented on augur/augur_client.py:
```diff
@@ -44,76 +44,105 @@ def __init__(self, hostname, port):
+      response_data['lastTradeBlockNumber']
+      if response_data['lastTradeBlockNumber'] else None)
+    # Ensure all None are omitted from tags.
+    response_data['tags'] = [tag for tag in response_data['tags'] if tag]
```
Can ```response_data['tags']``` equal ```None```? If so this will raise an exception.

kranar commented on augur/augur_client.py:
```diff
@@ -44,76 +44,105 @@ def __init__(self, hostname, port):
+      if response_data['resolutionSource'] else '')
+    # Cast NormalizedPayout types.
+    if not response_data['consensus']:
+      response_data['consensus'] = list()
```
Prefer ```[]``` over ```list()```

kranar commented on augur/augur_client.py:
```diff
@@ -161,3 +190,6 @@ def _require_is_open(self):
     return re.sub('([a-z0-9])([A-Z])', r'\1_\2', s1).lower()
+
+  def _ensure_unix_timestamp(self, timestamp):
+    return datetime.fromtimestamp(int(str(timestamp)[:10]))
```
What is the format used by Ethereum? Can you look into parsing it using ```datetime.strptime```?

joelaro04 replied:
Ethereum should be storing timestamps in seconds, however, when I ran `load_market_info` through all markets I got some weird timestamps like:
```
1535141988837623
91268587776
91268587776
91268587776
91302002688
91302002688
91302002688
```

joelaro04 replied:
It seems like these only appear on the Rinkeby test network. The first timestamp is NaN, the others are set to a far off date in the future. Running it on the Mainnet didn't produce any of these errors. Would it be safe to assume that the timestamp passed to `load_market_info` will be in the correct format?

kranar reviewed (approved):
Big improvement.

kranar commented:
Great, can you review what those timestamps represent?

On Fri, Dec 7, 2018 at 1:42 PM Joel Aro <notifications@github.com> wrote:

> *@joelaro04* commented on this pull request.
> ------------------------------
>
> In augur/augur_client.py
> <https://github.com/spiretrading/pyaugur/pull/2#discussion_r239905794>:
>
> > @@ -161,3 +190,6 @@ def _require_is_open(self):
>    def _to_snake_case(self, text):
>      s1 = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', text)
>      return re.sub('([a-z0-9])([A-Z])', r'\1_\2', s1).lower()
> +
> +  def _ensure_unix_timestamp(self, timestamp):
> +    return datetime.fromtimestamp(int(str(timestamp)[:10]))
>
> Ethereum should be storing timestamps in seconds, however, when I ran
> load_market_info through all markets I got some weird timestamps like:
>
> 1535141988837623
> 91268587776
> 91268587776
> 91268587776
> 91302002688
> 91302002688
> 91302002688
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/spiretrading/pyaugur/pull/2#discussion_r239905794>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APzFGZCmJ4NpaJRums2Cng7ge1FhCd-cks5u2rZ4gaJpZM4ZE9zy>
> .
>


-- 
Kamal Mansouri
Computer Programmer
+1 647-648-0498

kranar commented:
Good catch, for NaN convert it to the maximum supported datetime.datetime.
If a date is far off in the future that is fine, keep it.

On Fri, Dec 7, 2018 at 3:38 PM Joel Aro <notifications@github.com> wrote:

> *@joelaro04* commented on this pull request.
> ------------------------------
>
> In augur/augur_client.py
> <https://github.com/spiretrading/pyaugur/pull/2#discussion_r239939666>:
>
> > @@ -161,3 +190,6 @@ def _require_is_open(self):
>    def _to_snake_case(self, text):
>      s1 = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', text)
>      return re.sub('([a-z0-9])([A-Z])', r'\1_\2', s1).lower()
> +
> +  def _ensure_unix_timestamp(self, timestamp):
> +    return datetime.fromtimestamp(int(str(timestamp)[:10]))
>
> It seems like these only appear on the Rinkeby test network. The first
> timestamp is NaN, the others are set to a far off date in the future.
> Running it on the Mainnet didn't produce any of these errors. Would it be
> safe to assume that the timestamp passed to load_market_info will be in
> the correct format?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/spiretrading/pyaugur/pull/2#discussion_r239939666>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APzFGYOObD211AjCbzk94m9chIUcAau7ks5u2tHagaJpZM4ZE9zy>
> .
>


-- 
Kamal Mansouri
Computer Programmer
+1 647-648-0498

joelaro04 committed:
Changed list() to []. Added exception for timestamps.

joelaro04 committed:
Removed get_markets for testing

kranar commented on augur/augur_client.py:
```diff
@@ -57,17 +57,20 @@ def __init__(self, hostname, port):
+    if response_data['initialReportSize']:
+      response_data['initialReportSize'] = (
+        Decimal(response_data['initialReportSize']))
+    if response_data['finalizationBlockNumber']:
```
The API states that some of these are ```int```, not ```Decimal```. Review every single parameter very carefully to ensure it's initialized properly.

joelaro04 committed:
Update comments.

kranar commented:
There appears to be a lot of changes that are still needed to ensure the correctness of this case. Very thoroughly go over your implementation line by line, resolve any issue, and then submit a PR when you are confident you have properly validated all data.
